Related origins for passkeys

aznair opened the pull request:
#### ff54b35d1761a31af7ee619b0ad2b9e9bfea3e72
<pre>
Related origins for passkeys
<a href="https://bugs.webkit.org/show_bug.cgi?id=268426">https://bugs.webkit.org/show_bug.cgi?id=268426</a>
<a href="https://rdar.apple.com/121477240">rdar://121477240</a>

Reviewed by Pascoe.

This patch implements the WebKit side of related origins for passkeys.

- Validating the RPID against the current origin is now done by AuthenticationServices.

* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
(WebCore::AuthenticatorAssertionResponse::relyingPartyIdentifier const):
(WebCore::AuthenticatorAssertionResponse::setRelyingPartyIdentifier):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
(NS_ERROR_ENUM):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APIWebAuthenticationAssertionResponse.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKPublicKeyCredentialRequestOptions.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationAssertionResponse.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationAssertionResponse.mm:
(-[_WKWebAuthenticationAssertionResponse relyingPartyIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(relatedOrigins):
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::getExistingCredentials):
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::getAssertion):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::constructASController):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:

Canonical link: <a href="https://commits.webkit.org/274592@main">https://commits.webkit.org/274592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea6df9247270f6ea0600f0a8262fd4594e3bf3ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39449 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41983 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35349 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41755 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15757 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32974 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34167 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13483 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13458 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43261 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35819 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35443 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39250 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37500 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15867 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34391 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8847 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15915 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -297,11 +312,15 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
     m_completionHandler = WTFMove(handler);
     m_delegate = adoptNS([[_WKASDelegate alloc] initWithPage:WTFMove(requestData.page) completionHandler:makeBlockPtr([this](ASAuthorization *auth, NSError *error) mutable {
-        ensureOnMainRunLoop([this, auth = retainPtr(auth)]() {
+        ensureOnMainRunLoop([this, auth = retainPtr(auth), error]() {
```
error = retainPtr(error)

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -166,17 +166,18 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
 }
 
-RetainPtr<NSArray> WebAuthenticatorCoordinatorProxy::requestsForRegisteration(const PublicKeyCredentialCreationOptions &options, const Vector<uint8_t> &hash)
+RetainPtr<NSArray> WebAuthenticatorCoordinatorProxy::requestsForRegisteration(const PublicKeyCredentialCreationOptions &options, const Vector<uint8_t> &hash, const String &origin)
```
& on wrong side, for all of them.

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -196,7 +197,14 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
     if (includePlatformRequest) {
-        auto request = adoptNS([[allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:*options.rp.id] createCredentialRegistrationRequestWithChallenge:toNSData(options.challenge).get() name:options.user.name userID:toNSData(options.user.id).get()]);
+        RetainPtr<ASAuthorizationPlatformPublicKeyCredentialRegistrationRequest> request;
+        ASAuthorizationPlatformPublicKeyCredentialProvider *provider = [allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:*options.rp.id];
```
Should use a RetainPtr and a `adoptNS()` to avoid leaking.

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -196,7 +197,14 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
+        RetainPtr<ASAuthorizationPlatformPublicKeyCredentialRegistrationRequest> request;
+        ASAuthorizationPlatformPublicKeyCredentialProvider *provider = [allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:*options.rp.id];
+        if ([provider respondsToSelector:@selector(createCredentialRegistrationRequestWithClientData:name:userID:)]) {
+            ASPublicKeyCredentialClientData *clientData = [allocASPublicKeyCredentialClientDataInstance() initWithChallenge:toNSData(options.challenge).get() origin:origin];
```
Should use a RetainPtr and a adoptNS() to avoid leaking.

aznair replied:
Done

cdumez replied:
This doesn't look done? Still no RetainPtr or adoptNS().

aznair replied:
I don't know what happened, I swear I wasn't just lying :(
It should really be done this time.

cdumez replied:
No worries, it happens.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -219,7 +227,7 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
 }
 
-RetainPtr<NSArray> WebAuthenticatorCoordinatorProxy::requestsForAssertion(const PublicKeyCredentialRequestOptions &options, const Vector<uint8_t> &hash, std::optional<WebCore::SecurityOriginData> &parentOrigin)
+RetainPtr<NSArray> WebAuthenticatorCoordinatorProxy::requestsForAssertion(const PublicKeyCredentialRequestOptions &options, const Vector<uint8_t> &hash, std::optional<WebCore::SecurityOriginData> &parentOrigin, const String& origin)
```
Some of them have the `&` on the wrong side.

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -237,7 +245,14 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
     if ([platformAllowedCredentials count] || ![crossPlatformAllowedCredentials count]) {
-        auto request = adoptNS([[allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:options.rpId] createCredentialAssertionRequestWithChallenge:toNSData(options.challenge).get()]);
+        RetainPtr<ASAuthorizationPlatformPublicKeyCredentialAssertionRequest> request;
+        ASAuthorizationPlatformPublicKeyCredentialProvider *provider = [allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:options.rpId];
```
Should use a RetainPtr and a adoptNS() to avoid leaking.

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -237,7 +245,14 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
+        RetainPtr<ASAuthorizationPlatformPublicKeyCredentialAssertionRequest> request;
+        ASAuthorizationPlatformPublicKeyCredentialProvider *provider = [allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:options.rpId];
+        if ([provider respondsToSelector:@selector(createCredentialAssertionRequestWithClientData:)]) {
+            ASPublicKeyCredentialClientData *clientData = [allocASPublicKeyCredentialClientDataInstance() initWithChallenge:toNSData(options.challenge).get() origin:origin];
```
Should use a RetainPtr and a adoptNS() to avoid leaking.

aznair replied:
Done

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -297,11 +312,15 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
     m_completionHandler = WTFMove(handler);
     m_delegate = adoptNS([[_WKASDelegate alloc] initWithPage:WTFMove(requestData.page) completionHandler:makeBlockPtr([this](ASAuthorization *auth, NSError *error) mutable {
-        ensureOnMainRunLoop([this, auth = retainPtr(auth)]() {
+        ensureOnMainRunLoop([this, auth = retainPtr(auth), retainPtr(error)]() {
```
Not new to this patch but what guarantees that `this` is still alive when the lambda runs asynchronously? Odds are that we want to capture a WeakPtr instead.

aznair replied:
Since I don't know this code that well, I'm going to let @pascoej comment on this before making changes.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
```diff
@@ -105,8 +105,8 @@ class WebAuthenticatorCoordinatorProxy : public IPC::MessageReceiver {
     RetainPtr<ASAuthorizationController> constructASController(WebAuthenticationRequestData&&);
-    RetainPtr<NSArray> requestsForRegisteration(const WebCore::PublicKeyCredentialCreationOptions&, const Vector<uint8_t>& hash);
-    RetainPtr<NSArray> requestsForAssertion(const WebCore::PublicKeyCredentialRequestOptions&, const Vector<uint8_t>& hash, std::optional<WebCore::SecurityOriginData>& parentOrigin);
+    RetainPtr<NSArray> requestsForRegisteration(const WebCore::PublicKeyCredentialCreationOptions&, const Vector<uint8_t>& hash, const String&);
```
Typo: requestsForRegisteration
Also, please don't omit the new parameter's name since it is not obvious what it is from its type.

aznair replied:
Done (omitting the name was not intentional)

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -297,11 +312,15 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
     m_completionHandler = WTFMove(handler);
     m_delegate = adoptNS([[_WKASDelegate alloc] initWithPage:WTFMove(requestData.page) completionHandler:makeBlockPtr([this](ASAuthorization *auth, NSError *error) mutable {
-        ensureOnMainRunLoop([this, auth = retainPtr(auth)]() {
+        ensureOnMainRunLoop([this, auth = retainPtr(auth), error = retainPtr(error)]() {
```
@pascoej Shouldn't we capture a weakThis here? How can we guarantee that `this` is still alive when the lambda runs asynchronously?

aznair replied:
Changed to weakThis

pascoej commented on LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https.html:
```diff
@@ -29,69 +29,6 @@
             navigator.credentials.create(options), "Operation timed out.");
     }, "PublicKeyCredential's [[create]] with timeout");
 
-    promise_test(function(t) {
```
This is sad but I understand why. I hope we can readd these with some new test development.

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
```diff
@@ -105,8 +105,8 @@ class WebAuthenticatorCoordinatorProxy : public IPC::MessageReceiver {
     RetainPtr<ASAuthorizationController> constructASController(WebAuthenticationRequestData&&);
-    RetainPtr<NSArray> requestsForRegisteration(const WebCore::PublicKeyCredentialCreationOptions&, const Vector<uint8_t>& hash);
-    RetainPtr<NSArray> requestsForAssertion(const WebCore::PublicKeyCredentialRequestOptions&, const Vector<uint8_t>& hash, std::optional<WebCore::SecurityOriginData>& parentOrigin);
+    RetainPtr<NSArray> requestsForRegistration(const WebCore::PublicKeyCredentialCreationOptions&, const Vector<uint8_t>& hash, const String& origin);
```
These are going to conflict with https://github.com/WebKit/WebKit/pull/23750

aznair replied:
Yeah I'll rebase once that's landed. I think that will actually remove the need for most of this PR.

aznair replied:
Rebased.

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -237,7 +245,14 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
+            RetainPtr<ASPublicKeyCredentialClientData> clientData = adoptNS([allocASPublicKeyCredentialClientDataInstance() initWithChallenge:toNSData(options.challenge).get() origin:origin]);
+            request = adoptNS([provider.get() createCredentialAssertionRequestWithClientData:clientData.get()]);
+        } else
+            request = adoptNS([provider.get() createCredentialAssertionRequestWithChallenge:toNSData(options.challenge).get()]);
```
nit: You don't need to use .get() whenever something is on the receiver side of a obj-c style call. i.e. [provider.get() ...] can be [provider ...].

jasl23 committed:
Related origins for passkeys
https://bugs.webkit.org/show_bug.cgi?id=268426
rdar://121477240

Reviewed by Pascoe.

This patch implements the WebKit side of related origins for passkeys.

- Validating the RPID against the current origin is now done by AuthenticationServices.

* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
(WebCore::AuthenticatorAssertionResponse::relyingPartyIdentifier const):
(WebCore::AuthenticatorAssertionResponse::setRelyingPartyIdentifier):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredentialRequestOptions.idl:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
(NS_ERROR_ENUM):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APIWebAuthenticationAssertionResponse.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKPublicKeyCredentialRequestOptions.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationAssertionResponse.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationAssertionResponse.mm:
(-[_WKWebAuthenticationAssertionResponse relyingPartyIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(relatedOrigins):
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticatorInternal::getExistingCredentials):
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::getAssertion):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::constructASController):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:

Canonical link: https://commits.webkit.org/274592@main
